src/ohos/mod.rs

```diff
@@ -0,0 +1,271 @@
+
+static COUNTER: Counter = Counter::new();
+const MAIN_PIPE_TIMEOUT: Duration = Duration::from_secs(10);
+
+pub struct InnerWebView {
+  id: String,
+  webview: Webview,
+}
+
+impl InnerWebView {
+  pub fn new_as_child(
+    _window: &impl HasWindowHandle,
+    attributes: WebViewAttributes,
+    pl_attrs: super::PlatformSpecificWebViewAttributes,
+  ) -> Result<Self> {
+    Self::new(_window, attributes, pl_attrs)
+  }
+
+  pub fn new(
+    _window: &impl HasWindowHandle,
+    attributes: WebViewAttributes,
+    _pl_attrs: super::PlatformSpecificWebViewAttributes,
+  ) -> Result<Self> {
+    let WebViewAttributes {
+      id,
+      url,
+      html,
+      initialization_scripts,
+      ipc_handler,
+      #[cfg(any(debug_assertions, feature = "devtools"))]
+      devtools,
+      custom_protocols,
+      background_color,
+      transparent,
+      headers,
+      autoplay,
+      user_agent,
+      javascript_disabled,
+      ..
+    } = attributes;
+
+    let u = url.clone().unwrap_or_default();
+
+    let id = id
+      .map(|id| id.to_string())
+      .unwrap_or_else(|| COUNTER.next().to_string());
+
+    let background_color =
+      background_color.map(|c| format!("#{:02X}{:02X}{:02X}{:02X}", c.0, c.1, c.2, c.3));
+
+    let mut webview_builder = WebViewBuilder::new()
+      .id(id.clone())
+      .style(WebViewStyle {
+        x: None,
+        y: None,
+        visible: None,
+        background_color,
+      })
+      .url(u)
+      .javascript_enabled(!javascript_disabled)
+      .autoplay(autoplay)
+      .initialization_scripts(vec![initialization_scripts
+        .iter()
+        .map(|s| s.script.clone())
+        .collect::<Vec<_>>()
+        .join("\n")])
+      .devtools(devtools)
+      .transparent(transparent);
+
+    if let Some(html) = html {
+      webview_builder = webview_builder.html(html);
+    }
+
+    if let Some(headers) = headers {
+      webview_builder = webview_builder.headers(headers);
+    }
+
+    if let Some(user_agent) = user_agent {
+      webview_builder = webview_builder.user_agent(user_agent);
+    }
+
+    let webview = webview_builder
+      .build()
+      .map_err(|e| Error::OpenHarmonyInitError(e.to_string()))?;
+
+    let current_id = id.clone();
+    webview
+      .on_controller_attach(move || {
+        let _builder = WebProxyBuilder::new(current_id.clone(), "ipc".to_string())
+          .add_method("postMessage", |_, params| {
+            if let Some(ipc_handler) = &ipc_handler {
+              let message = params.get(0).unwrap_or(&"".to_string()).to_owned();
+              ipc_handler(Request::new(message));
+            }
+          })
+          .build()
+          .expect("Failed to build web proxy");
+      })
+      .map_err(|e| {
+        Error::OpenHarmonyInitError(format!("Failed to add controller attach listener: {}", e))
+      })?;
+
+    for (protocol, callback) in custom_protocols {
+      webview
+        .custom_protocol(protocol, move |web, req, _flag| {
```

Comment:
async implementation needs https://github.com/harmony-contrib/openharmony-ability/pull/29
we need that to merge this, otherwise some tauri commands will be blocking and have this timeout set here